lib/linters.py

```diff
@@ -0,0 +1,612 @@
+
+
+def sanitizeNetworkItem(x):
+  # Remove comments and identify just the content we care about
+  return x.split('#', 1)[0].strip()
+
+
+class severity(object):
+  DISABLED = 'disabled'
+  ADVICE = 'advice'
+  WARNING = 'warning'
+  AUTOFIX = 'autofix'
+  ERROR = 'error'
+
+  @classmethod
+  def human(cls, k):
+    mapping = {
+      cls.DISABLED: 'Disabled',
+      cls.ADVICE: 'Advice',
+      cls.WARNING: 'Warning',
+      cls.ERROR: 'Error',
+      cls.AUTOFIX: 'Auto-Fix',
+    }
+    return mapping[k]
+
+  @classmethod
+  def int(cls, k):
+    mapping = {
+      cls.DISABLED: 10,
+      cls.ADVICE: 20,
+      cls.WARNING: 30,
+      cls.ERROR: 40,
+      cls.AUTOFIX: 25,
+    }
+    return mapping[k]
+
+
+class LintErrors(object):
+  def __init__(self):
+    self.errs = []
+    self.filename = None
+
+  def __iter__(self):
+    for e in self.errs:
+      yield e
+
+  def add(self, severity, msg, lineno=0, offset=0, filename=None,
+      category='CAP'):
+    fn = filename or self.filename or 'unknown'
+    self.errs.append(Error(fn, lineno, offset, severity, msg, category))
+
+  def tree(self, sort=True):
+    t = defaultdict(list)
+    for e in self.errs:
+      t[e.filename].append(e)
+    if sort:
+      for v in t.values():
+        v.sort(key=lambda e: e.lineno)
+    return t
+
+  def pprint(self):
+    for filename, errors in self.tree().iteritems():
+      print("%s -" % filename)
+      for error in errors:
+        print("  %5d <%s> %s" % (error.lineno,
+                     severity.human(error.severity),
+                     error.message))
+
+  def plain(self):
+    for error in sorted(self.errs, key=lambda x: (x.filename, x.lineno)):
+      print(error)
+
+  def json(self):
+    print(json.dumps(self.tree()))
+
+
+def register_linter(f):
+  """Class decorator to register a linter to be executed."""
+  globals().setdefault('CAPIRCA_LINTERS', []).append(f)
+  return f
+
+
+def get_linters():
+  return globals().get('CAPIRCA_LINTERS', [])
+
+
+def build_linters(configpath):
+  errors = LintErrors()
+  if configpath:
+    with open(configpath) as f:
+      config = yaml.load(f.read())
+  else:
+    config = {}
+  linters = [kls(errors=errors, config=config) for kls in get_linters()]
+  return (errors, linters)
+
+
+class BaseLintRule(object):
+
+  """Base Lint Rule class. All lint rules should extend this.
+
+  All configurable parameters should be placed into DEFAULTS - as these
+  values get merged with the optional user-specified YAML file.
+  """
+
+  CONFIG_NAME = None
+  DEFAULT_ENABLED = True
+  CATEGORY = None
+  DEFAULTS = {}
+
+  def __init__(self, errors, config):
+    self.global_config = config
+    self.errors = errors
+    # build a local copy of the linter specific config
+    cn = self.CONFIG_NAME or self.__class__.__name__
+    local_config = deepcopy(self.DEFAULTS)
+    local_config.update(self.global_config.get(cn, {}))
+    self.config = local_config
+    self.setup()
+
+  def setup(self):
```

Comment:
What is the difference between setup and init? When would I use setup instead of just adding more things to init? Maybe add that to the docstring?